<a href="https://colab.research.google.com/github/nrios256-create/analisis_datos/blob/main/datos_crudos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
import pandas as pd
import numpy as np
#importo librerias a utilizar

In [7]:
datos_cr = pd.read_csv("datos_crudos_tarea.csv") #cargo la base de datos
datos_cr.head() #reviso que si sea la base de datos correcta

,Unnamed: 0,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,presion_arterial,glucosa_mg_dL,colesterol_total,fuma,actividad_fisica,imc_calculado
0,0,524,2022-01-23 00:00:00,73,MUJER,80.6,2.38,89/97,NaN,-100,0,alta,NaN
1,1,603,2023-03-02 00:00:00,7,X,126.3,1.18,88/79,95,-100,0,Sedentario,NaN
2,2,527,2022-04-28 00:00:00,68,F,37.7,20,118/51,500,-100,no,Sedentario,NaN
3,3,32,2022-05-06 00:00:00,96,f,47.5,0.45,195/105,300,bueno,SI,Sedentario,NaN
4,4,617,2023-06-07 00:00:00,72,Mujer,198.8,0.81,126/87,85,120,0,Baja,NaN


In [8]:
datos_cr.info()
datos_cr.shape
#reviso la informacion de cada columna

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        1020 non-null   int64  
 1   id_paciente       1020 non-null   int64  
 2   fecha_consulta    1020 non-null   object 
 3   edad              1006 non-null   object 
 4   sexo              917 non-null    object 
 5   peso_kg           1006 non-null   object 
 6   altura_m          1006 non-null   object 
 7   presion_arterial  1020 non-null   object 
 8   glucosa_mg_dL     923 non-null    object 
 9   colesterol_total  886 non-null    object 
 10  fuma              908 non-null    object 
 11  actividad_fisica  929 non-null    object 
 12  imc_calculado     0 non-null      float64
dtypes: float64(1), int64(2), object(10)
memory usage: 103.7+ KB


(1020, 13)

In [9]:
datos_cr.columns #consulto el nombre de las columnas para tenerlas claras y ver si hay errores para luego corregirlos

Index(['Unnamed: 0', 'id_paciente', 'fecha_consulta', 'edad', 'sexo',
       'peso_kg', 'altura_m', 'presion_arterial', 'glucosa_mg_dL',
       'colesterol_total', 'fuma', 'actividad_fisica', 'imc_calculado'],
      dtype='object')

Revisé la base y pues sí tiene varios datos faltantes en columnas como sexo, glucosa y colesterol. también miré si había duplicados y no hay y vi que hay una columna que se llama "unnamed: 0" que básicamente no sirve para nada, parece que es solo un índice que se guardó ahí.

In [10]:
datos_cr = datos_cr.drop(columns=['Unnamed: 0'])  # uso drop para eliminar esa columna porque no sirve, es solo un índice que se guardó ahí y no aporta nada

In [11]:
datos_cr.columns #aqui confirmo que ya no este la columna unnamed:0

Index(['id_paciente', 'fecha_consulta', 'edad', 'sexo', 'peso_kg', 'altura_m',
       'presion_arterial', 'glucosa_mg_dL', 'colesterol_total', 'fuma',
       'actividad_fisica', 'imc_calculado'],
      dtype='object')

In [12]:
datos_cr.head(20) #aqui reviso como se ve el dataset sin esa columna

,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,presion_arterial,glucosa_mg_dL,colesterol_total,fuma,actividad_fisica,imc_calculado
0,524,2022-01-23 00:00:00,73,MUJER,80.6,2.38,89/97,NaN,-100,0,alta,NaN
1,603,2023-03-02 00:00:00,7,X,126.3,1.18,88/79,95,-100,0,Sedentario,NaN
2,527,2022-04-28 00:00:00,68,F,37.7,20,118/51,500,-100,no,Sedentario,NaN
3,32,2022-05-06 00:00:00,96,f,47.5,0.45,195/105,300,bueno,SI,Sedentario,NaN
4,617,2023-06-07 00:00:00,72,Mujer,198.8,0.81,126/87,85,120,0,Baja,NaN
5,586,2022-12-28 00:00:00,48,f,104.7,1.11,167/106,bajo,350,no,0,NaN
6,445,2023-11-17 00:00:00,37,Mujer,18.2,2.17,212/73,500,120,Fumador,baja,NaN
7,733,2021-12-09 00:00:00,55,Mujer,92.7,1.47,109/119,110,200,Fumador,Alta,NaN
8,77,2021-02-05 00:00:00,72,m,93.9,0.82,104/110,95,NaN,NO,NaN,NaN
9,412,2022-09-29 00:00:00,13,Hombre,73.4,2.47,88/116,180,-100,Fumador,1,NaN


In [13]:
datos_cr.columns = datos_cr.columns.str.strip().str.lower().str.replace(' ', '_') # # arreglo los nombres de las columnas quitando espacios, poniendo todo en minúscula y usando guion bajo para que sea más ordenado
#str.strip() = quita espacios al inicio y final
#str.lower() = pone todo en minúsculas
#str.replace(' ', '_') = cambia espacios por guion bajo
datos_cr.columns

Index(['id_paciente', 'fecha_consulta', 'edad', 'sexo', 'peso_kg', 'altura_m',
       'presion_arterial', 'glucosa_mg_dl', 'colesterol_total', 'fuma',
       'actividad_fisica', 'imc_calculado'],
      dtype='object')

In [14]:
datos_cr.duplicated().sum() #revisar si hay duplicados

np.int64(19)

In [15]:
datos_cr[datos_cr.duplicated()] #aqui le pido que me muestre solo las filas que estan duplicadas

,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,presion_arterial,glucosa_mg_dl,colesterol_total,fuma,actividad_fisica,imc_calculado
136,137,2021-02-01 00:00:00,85,Hombre,-38.5,2.13,170/76,85,malo,Sí,Sedentario,NaN
278,762,2022-04-02 00:00:00,69,Mujer,-34.4,0,124/51,-20,NaN,Sí,alta,NaN
410,812,2023-12-16 00:00:00,85,H,-18.8,0.41,142/68,500,bueno,NaN,media,NaN
434,939,2023-11-18 00:00:00,75,f,-48.2,0.72,139/100,alto,malo,0,0,NaN
445,679,2022-06-04 00:00:00,66,H,42.9,0.7,106/101,-20,350,NaN,baja,NaN
517,900,2020-10-29 00:00:00,-33,f,111.5,2.44,212/baja,180,600,Fumador,1,NaN
529,661,2020-12-13 00:00:00,noventa,M,1000,0.88,87/109,-20,120,0,2,NaN
559,77,2021-02-05 00:00:00,72,m,93.9,0.82,104/110,95,NaN,NO,NaN,NaN
573,320,2023-07-24 00:00:00,desconocido,Mujer,136.5,1.85,198/88,110,-100,NO,Baja,NaN
633,412,2022-09-29 00:00:00,13,Hombre,73.4,2.47,88/116,180,-100,Fumador,1,NaN


In [16]:
datos_cr[datos_cr.duplicated(keep=False)].sort_values(by='id_paciente')
# aquí le estoy pidiendo que me muestre todas las filas que están duplicadas, no solo una parte,
# por eso uso keep=False para ver tanto la original como la repetida,
# y luego las ordeno por id_paciente con sort_values para poder verlas mejor organizadas

,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,presion_arterial,glucosa_mg_dl,colesterol_total,fuma,actividad_fisica,imc_calculado
8,77,2021-02-05 00:00:00,72,m,93.9,0.82,104/110,95,NaN,NO,NaN,NaN
559,77,2021-02-05 00:00:00,72,m,93.9,0.82,104/110,95,NaN,NO,NaN,NaN
136,137,2021-02-01 00:00:00,85,Hombre,-38.5,2.13,170/76,85,malo,Sí,Sedentario,NaN
129,137,2021-02-01 00:00:00,85,Hombre,-38.5,2.13,170/76,85,malo,Sí,Sedentario,NaN
148,281,2020-04-13 00:00:00,57,Mujer,5000,bajo,500/baja,180,350,no,Sedentario,NaN
916,281,2020-04-13 00:00:00,57,Mujer,5000,bajo,500/baja,180,350,no,Sedentario,NaN
573,320,2023-07-24 00:00:00,desconocido,Mujer,136.5,1.85,198/88,110,-100,NO,Baja,NaN
198,320,2023-07-24 00:00:00,desconocido,Mujer,136.5,1.85,198/88,110,-100,NO,Baja,NaN
633,412,2022-09-29 00:00:00,13,Hombre,73.4,2.47,88/116,180,-100,Fumador,1,NaN
9,412,2022-09-29 00:00:00,13,Hombre,73.4,2.47,88/116,180,-100,Fumador,1,NaN


In [17]:
datos_cr = datos_cr.drop_duplicates() #con este drop pido que elimine los duplicados
datos_cr.duplicated().sum() #aqui reviso que queden eliminados

np.int64(0)

In [18]:
datos_cr.isnull().sum()  # aqui reviso cuantos valores faltantes hay en cada columna

,0
id_paciente,0
fecha_consulta,0
edad,13
sexo,103
peso_kg,14
altura_m,14
presion_arterial,0
glucosa_mg_dl,97
colesterol_total,132
fuma,110


In [19]:
datos_cr['sexo'].unique()  # aqui reviso las diferentes formas en que está escrita la variable sexo

array(['MUJER', 'X', 'F', 'f', 'Mujer', 'm', 'Hombre', 'M', nan, 'H'],
      dtype=object)

In [20]:
datos_cr['sexo'] = datos_cr['sexo'].str.strip().str.upper()
# aqui limpio los textos quitando espacios con strip y poniendo todo en mayúscula con upper

datos_cr['sexo'] = datos_cr['sexo'].replace({
    'MUJER': 'F',
    'F': 'F',
    'HOMBRE': 'M',
    'M': 'M',
    'H': 'M',
    'X': np.nan
})
# aqui unifico las categorias para que solo queden F y M con replace, y convierto valores raros como X en nulos con np.nan como enseño el profe
datos_cr['sexo'].unique() #reviso que si se aplique lo anterior

array(['F', nan, 'M'], dtype=object)

In [21]:
datos_cr['fuma'].unique()  #reviso como están escritos los valores de los fumadores

array(['0', 'no', 'SI', 'Fumador', 'NO', 'No', '1', 'Sí', nan, 'si'],
      dtype=object)

In [35]:
datos_cr['fuma'] = datos_cr['fuma'].str.strip().str.title()
# aqui limpio los textos quitando espacios con strip y poniendo primera en mayuscula y demas en minuscula

datos_cr['fuma'] = datos_cr['fuma'].replace({
    'SI': 'SI',
    'SÍ': 'SI',
    'Fumador': 'SI',
    'NO': 'NO',
    '1': np.nan,
    '0': np.nan
})
# aqui unifico las formas de si y no utilizando replace, y en el caso de 1 y 0 los paso a nulos porque pues el profe no hizo aclaracion al respecto
#de como tratar 1 y 0 y no tiene mucho sentido que pongan valores numericos asi que los dejare como valores nulos por la falta de informacion

datos_cr['fuma'].unique()
# aqui reviso que ya quede limpio y solo tenga SI, NO o nulos(nan)

array([nan, 'No', 'Si', 'SI', 'Sí'], dtype=object)

In [23]:
datos_cr['actividad_fisica'].unique()  # aqui reviso como están escritos

array(['alta', 'Sedentario', 'Baja', '0', 'baja', 'Alta', nan, '1',
       'Media', '2', 'media'], dtype=object)

In [37]:
datos_cr['actividad_fisica'] = datos_cr['actividad_fisica'].str.strip().str.title()
# aqui limpio los textos quitando espacios con strip y poniendo primera en mayuscula y demas en minuscula

datos_cr['actividad_fisica'] = datos_cr['actividad_fisica'].replace({
    'Alta': 'Alta',
    'Media': 'Media',
    'Baja': 'Baja',
    'Sedentario': 'Sedentario',
    '0': np.nan,
    '1': np.nan,
    '2': np.nan
})
# aqui unifico las categorias para que solo queden niveles claros de actividad fisica,
# y los valores 0, 1 y 2 los paso a nulos porque no es claro que representan y pues aqui no es logico poner numeros

#en la variable “sedentario” decidí dejarla como una categoría aparte, porque aunque podría interpretarse como baja actividad,
#no es exactamente lo mismo porque tambien podria ser que no hace nada de ejercicio y preferí no asumir cosas que no están claras en los datos.

datos_cr['actividad_fisica'].unique()
# aqui reviso que ya quede limpio

array(['Alta', 'Sedentario', 'Baja', nan, 'Media'], dtype=object)

In [25]:
datos_cr.dtypes #ver qué tipo de dato tiene cada columna y asi saber cuales estan mal y necesito convertir

,0
id_paciente,int64
fecha_consulta,object
edad,object
sexo,object
peso_kg,object
altura_m,object
presion_arterial,object
glucosa_mg_dl,object
colesterol_total,object
fuma,object


revisé los tipos de datos de cada columna y me di cuenta que varias variables numéricas como edad, peso, altura, glucosa y colesterol están como texto (object), por lo que es necesario convertirlas a numéricas. también la fecha está como texto y debería convertirse a formato de fecha.

In [26]:
datos_cr['edad'] = pd.to_numeric(datos_cr['edad'], errors='coerce')
datos_cr['peso_kg'] = pd.to_numeric(datos_cr['peso_kg'], errors='coerce')
datos_cr['altura_m'] = pd.to_numeric(datos_cr['altura_m'], errors='coerce')
datos_cr['glucosa_mg_dl'] = pd.to_numeric(datos_cr['glucosa_mg_dl'], errors='coerce')
datos_cr['colesterol_total'] = pd.to_numeric(datos_cr['colesterol_total'], errors='coerce')
# aqui convierto cada variable de texto que este mal a número con pd.to_numeric,
#y uso errors='coerce' para que los valores raros o mal escritos se vuelvan nulos

In [27]:
datos_cr[['edad', 'peso_kg', 'altura_m', 'glucosa_mg_dl', 'colesterol_total']].describe()
#después de convertir las variables a numéricas, revisé sus estadísticas para identificar posibles valores fuera de rango o inconsistentes en los datos

,edad,peso_kg,altura_m,glucosa_mg_dl,colesterol_total
count,950.000000,947.000000,939.000000,719.000000,601.000000
mean,65.204211,399.878141,1.782758,168.956885,237.154742
std,117.890183,1576.462972,3.308387,157.342651,238.175127
min,-49.000000,-99.700000,-2.450000,-20.000000,-100.000000
25%,22.000000,39.050000,0.740000,67.500000,120.000000
50%,51.000000,94.400000,1.360000,110.000000,200.000000
75%,84.000000,151.500000,1.990000,300.000000,350.000000
max,999.000000,9999.000000,20.000000,500.000000,600.000000


al revisar las estadísticas encontré varios valores fuera de rango en diferentes variables. por ejemplo, hay edades negativas y valores muy altos como 999, pesos negativos o exagerados, alturas imposibles y valores negativos en glucosa y colesterol. Esto indica que hay errores en los datos que deben ser tratados.

In [28]:
# EDAD
datos_cr.loc[(datos_cr['edad'] < 0) | (datos_cr['edad'] > 100), 'edad'] = np.nan
# aqui uso loc para ubicar filas especificas que cumplen una condición, en este caso edades negativas o muy altas
# uso los paréntesis con condiciones y el operador | que significa "o" para filtrar esos valores
# luego en esas mismas filas reemplazo la columna edad por np.nan, que significa un valor nulo
# hago esto porque esas edades no tienen sentido en la vida real
#basicamente le digo que solo tome valores en edad mayor a 0 y menor a 100

# PESO
datos_cr.loc[(datos_cr['peso_kg'] < 30) | (datos_cr['peso_kg'] > 440), 'peso_kg'] = np.nan
# aqui hago lo mismo, uso loc para encontrar pesos muy bajos o exagerados

# ALTURA
datos_cr.loc[(datos_cr['altura_m'] < 1) | (datos_cr['altura_m'] > 2.72), 'altura_m'] = np.nan
## aqui filtro alturas que no son posibles usando loc

# GLUCOSA
datos_cr.loc[(datos_cr['glucosa_mg_dl'] < 0), 'glucosa_mg_dl'] = np.nan
# aqui uso loc para encontrar valores negativos en glucosa
# y los paso a np.nan porque no es lógico tener glucosa negativa

# COLESTEROL
datos_cr.loc[(datos_cr['colesterol_total'] < 0), 'colesterol_total'] = np.nan

#para todo lo demas basicamente es lo mismo pero con otras variables

In [29]:
datos_cr[['edad', 'peso_kg', 'altura_m', 'glucosa_mg_dl', 'colesterol_total']].describe()
# aqui vuelvo a revisar las estadísticas después de la limpieza para ver si ya se eliminaron los valores fuera de rango

,edad,peso_kg,altura_m,glucosa_mg_dl,colesterol_total
count,780.000000,687.000000,570.000000,633.000000,480.000000
mean,49.115385,114.113392,1.738912,194.628752,322.145833
std,29.027622,48.608998,0.438244,150.355790,187.359912
min,0.000000,30.500000,1.000000,50.000000,120.000000
25%,24.750000,67.800000,1.360000,85.000000,120.000000
50%,47.000000,116.200000,1.750000,110.000000,350.000000
75%,73.250000,155.100000,2.120000,300.000000,600.000000
max,100.000000,199.500000,2.500000,500.000000,600.000000


In [30]:
datos_cr.isnull().sum()
# aqui reviso si aumentaron los nulos, porque convertí valores inválidos en NaN

,0
id_paciente,0
fecha_consulta,0
edad,221
sexo,194
peso_kg,314
altura_m,431
presion_arterial,0
glucosa_mg_dl,368
colesterol_total,521
fuma,305


después de limpiar los valores fuera de rango, volví a revisar las estadísticas y ya los datos quedaron dentro de rangos más lógicos y realistas pues tuve que investigar cositas que no sabia para poder cuadrar los parametros reales. también aumentaron los valores nulos, lo cual tiene sentido porque muchos datos incorrectos se reemplazaron por NaN.

In [31]:
datos_cr['imc_calculado'] = datos_cr['peso_kg'] / (datos_cr['altura_m'] ** 2)

# aqui calculo el imc usando peso y altura, ya que esta columna estaba vacía

In [ ]:
datos_cr['fecha_consulta'] = pd.to_datetime(datos_cr['fecha_consulta'], errors='coerce')
# aqui convierto la columna fecha_consulta a formato fecha porque estaba como texto con pd.to_datetime
# uso errors='coerce' para que si llega a haber alguna fecha mal escrita se vuelva nula

In [32]:
datos_cr['fecha_consulta'].dtype
#reviso si ya quedó en formato fecha

dtype('O')

In [40]:
datos_cr.head(60)

,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,presion_arterial,glucosa_mg_dl,colesterol_total,fuma,actividad_fisica,imc_calculado
0,524,2022-01-23 00:00:00,73.0,F,80.6,2.38,89/97,NaN,NaN,NaN,Alta,14.229221
1,603,2023-03-02 00:00:00,7.0,NaN,126.3,1.18,88/79,95.0,NaN,NaN,Sedentario,90.706693
2,527,2022-04-28 00:00:00,68.0,F,37.7,NaN,118/51,500.0,NaN,No,Sedentario,NaN
3,32,2022-05-06 00:00:00,96.0,F,47.5,NaN,195/105,300.0,NaN,Si,Sedentario,NaN
4,617,2023-06-07 00:00:00,72.0,F,198.8,NaN,126/87,85.0,120.0,NaN,Baja,NaN
5,586,2022-12-28 00:00:00,48.0,F,104.7,1.11,167/106,NaN,350.0,No,NaN,84.976869
6,445,2023-11-17 00:00:00,37.0,F,NaN,2.17,212/73,500.0,120.0,SI,Baja,NaN
7,733,2021-12-09 00:00:00,55.0,F,92.7,1.47,109/119,110.0,200.0,SI,Alta,42.898792
8,77,2021-02-05 00:00:00,72.0,M,93.9,NaN,104/110,95.0,NaN,No,NaN,NaN
9,412,2022-09-29 00:00:00,13.0,M,73.4,2.47,88/116,180.0,NaN,SI,NaN,12.031012


In [42]:
datos_cr[['presion_arterial']].info()
#reviso tipo es presion_arterial

<class 'pandas.core.frame.DataFrame'>
Index: 1001 entries, 0 to 1019
Data columns (total 1 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   presion_arterial  1001 non-null   object
dtypes: object(1)
memory usage: 15.6+ KB


In [43]:
datos_cr[['presion_sistolica', 'presion_diastolica']] = datos_cr['presion_arterial'].str.split('/', expand=True)
# aqui separo la presion arterial en dos partes usando el slash como separador
# porque en esa columna vienen dos valores en uno solo
# uso split para separar el texto usando el "/" como referencia,
# y expand=True hace que cada parte se vaya a una columna diferente en lugar de quedar en una lista

In [44]:
datos_cr['presion_sistolica'] = pd.to_numeric(datos_cr['presion_sistolica'], errors='coerce')
datos_cr['presion_diastolica'] = pd.to_numeric(datos_cr['presion_diastolica'], errors='coerce')
# aqui convierto cada parte a número con pd.to_numeric
# uso errors='coerce' para que valores raros como "baja" se vuelvan nulos

In [47]:
datos_cr['presion_arterial'] = datos_cr['presion_sistolica'].astype(str) + '/' + datos_cr['presion_diastolica'].astype(str)
# aqui vuelvo a unir las dos columnas en una sola con formato numero/numero

In [51]:
datos_cr = datos_cr.drop(columns=['presion_arterial'])
# aqui elimino la columna original porque ya separé la información en dos variables más claras con drop

In [53]:
datos_cr.head(60)

,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,glucosa_mg_dl,colesterol_total,fuma,actividad_fisica,imc_calculado,presion_sistolica,presion_diastolica
0,524,2022-01-23 00:00:00,73.0,F,80.6,2.38,NaN,NaN,NaN,Alta,14.229221,89.0,97.0
1,603,2023-03-02 00:00:00,7.0,NaN,126.3,1.18,95.0,NaN,NaN,Sedentario,90.706693,88.0,79.0
2,527,2022-04-28 00:00:00,68.0,F,37.7,NaN,500.0,NaN,No,Sedentario,NaN,118.0,51.0
3,32,2022-05-06 00:00:00,96.0,F,47.5,NaN,300.0,NaN,Si,Sedentario,NaN,195.0,105.0
4,617,2023-06-07 00:00:00,72.0,F,198.8,NaN,85.0,120.0,NaN,Baja,NaN,126.0,87.0
5,586,2022-12-28 00:00:00,48.0,F,104.7,1.11,NaN,350.0,No,NaN,84.976869,167.0,106.0
6,445,2023-11-17 00:00:00,37.0,F,NaN,2.17,500.0,120.0,SI,Baja,NaN,212.0,73.0
7,733,2021-12-09 00:00:00,55.0,F,92.7,1.47,110.0,200.0,SI,Alta,42.898792,109.0,119.0
8,77,2021-02-05 00:00:00,72.0,M,93.9,NaN,95.0,NaN,No,NaN,NaN,104.0,110.0
9,412,2022-09-29 00:00:00,13.0,M,73.4,2.47,180.0,NaN,SI,NaN,12.031012,88.0,116.0


In [54]:
datos_cr.loc[
    (datos_cr['presion_sistolica'].isna()) |
    (datos_cr['presion_diastolica'].isna()) |
    (datos_cr['presion_sistolica'] < 70) |
    (datos_cr['presion_sistolica'] > 250) |
    (datos_cr['presion_diastolica'] < 40) |
    (datos_cr['presion_diastolica'] > 150),
    ['presion_sistolica', 'presion_diastolica']
] = np.nan
# aqui uso loc para buscar las filas donde alguna de las dos presiones esté nula o tenga valores fuera de rango
# con isna reviso si hay valores faltantes
# y con las demás condiciones reviso si los números son absurdos
# si algo está mal, pongo ambas columnas en nulo porque la presión arterial ya no sería confiable completa
#el | lo uso para decir que si se cumple una condición o la otra, entonces se haga algo

In [61]:
datos_cr[['presion_sistolica', 'presion_diastolica']].head(60)
# aqui reviso como quedaron las dos columnas después de la limpieza

,presion_sistolica,presion_diastolica
0,89.0,97.0
1,88.0,79.0
2,118.0,51.0
3,195.0,105.0
4,126.0,87.0
5,167.0,106.0
6,212.0,73.0
7,109.0,119.0
8,104.0,110.0
9,88.0,116.0


la presión arterial venía en una sola columna con dos valores, por lo que la separé en presión sistólica y diastólica. la columna original ya no era necesaria, así que la eliminé. entonces para los valores faltantes, decidí mantenerlos como nulos cuando alguna de las dos partes no era válida, para no asumir información incorrecta.

In [62]:
datos_cr.info()
# aqui reviso los tipos de datos finales y si quedaron como corresponde

<class 'pandas.core.frame.DataFrame'>
Index: 1001 entries, 0 to 1019
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_paciente         1001 non-null   int64  
 1   fecha_consulta      1001 non-null   object 
 2   edad                780 non-null    float64
 3   sexo                807 non-null    object 
 4   peso_kg             687 non-null    float64
 5   altura_m            570 non-null    float64
 6   glucosa_mg_dl       633 non-null    float64
 7   colesterol_total    480 non-null    float64
 8   fuma                696 non-null    object 
 9   actividad_fisica    634 non-null    object 
 10  imc_calculado       396 non-null    float64
 11  presion_sistolica   818 non-null    float64
 12  presion_diastolica  818 non-null    float64
dtypes: float64(8), int64(1), object(4)
memory usage: 109.5+ KB


In [63]:
datos_cr.isnull().sum()
# aqui reviso cuantos valores nulos quedaron al final

,0
id_paciente,0
fecha_consulta,0
edad,221
sexo,194
peso_kg,314
altura_m,431
glucosa_mg_dl,368
colesterol_total,521
fuma,305
actividad_fisica,367


In [64]:
datos_cr.head()
# aqui reviso visualmente como quedó la base

,id_paciente,fecha_consulta,edad,sexo,peso_kg,altura_m,glucosa_mg_dl,colesterol_total,fuma,actividad_fisica,imc_calculado,presion_sistolica,presion_diastolica
0,524,2022-01-23 00:00:00,73.0,F,80.6,2.38,NaN,NaN,NaN,Alta,14.229221,89.0,97.0
1,603,2023-03-02 00:00:00,7.0,NaN,126.3,1.18,95.0,NaN,NaN,Sedentario,90.706693,88.0,79.0
2,527,2022-04-28 00:00:00,68.0,F,37.7,NaN,500.0,NaN,No,Sedentario,NaN,118.0,51.0
3,32,2022-05-06 00:00:00,96.0,F,47.5,NaN,300.0,NaN,Si,Sedentario,NaN,195.0,105.0
4,617,2023-06-07 00:00:00,72.0,F,198.8,NaN,85.0,120.0,NaN,Baja,NaN,126.0,87.0


In [65]:
datos_cr.describe()
# aqui reviso las estadísticas para confirmar que ya no hay valores fuera de rango

,id_paciente,edad,peso_kg,altura_m,glucosa_mg_dl,colesterol_total,imc_calculado,presion_sistolica,presion_diastolica
count,1001.000000,780.000000,687.000000,570.000000,633.000000,480.000000,396.000000,818.000000,818.000000
mean,500.626374,49.115385,114.113392,1.738912,194.628752,322.145833,46.204080,149.799511,84.286064
std,288.702678,29.027622,48.608998,0.438244,150.355790,187.359912,34.953666,39.511942,20.254412
min,1.000000,0.000000,30.500000,1.000000,50.000000,120.000000,5.492475,80.000000,50.000000
25%,251.000000,24.750000,67.800000,1.360000,85.000000,120.000000,22.840108,117.000000,66.250000
50%,501.000000,47.000000,116.200000,1.750000,110.000000,350.000000,34.777011,151.000000,84.000000
75%,750.000000,73.250000,155.100000,2.120000,300.000000,600.000000,57.718653,184.000000,101.750000
max,1000.000000,100.000000,199.500000,2.500000,500.000000,600.000000,194.700000,219.000000,119.000000


In [68]:
datos_cr.loc[(datos_cr['imc_calculado'] < 10) | (datos_cr['imc_calculado'] > 60), 'imc_calculado'] = np.nan
# aqui uso loc para ubicar las filas donde el imc cumple ciertas condiciones
# uso (condicion1) | (condicion2) donde el | significa "o", es decir, valores menores a 10 o mayores a 60
# esos rangos no son realistas para el imc, por eso los considero errores
# y en esas filas reemplazo el valor por np.nan, que representa un valor nulo
# esto lo hago para eliminar datos que no son confiables sin borrar la fila completa
datos_cr[['imc_calculado']].describe()

,imc_calculado
count,289.000000
mean,31.903817
std,13.267664
min,10.138787
25%,21.355151
50%,30.848000
75%,40.320098
max,59.976380


In [71]:
datos_cr['fecha_consulta']= pd.to_datetime(datos_cr['fecha_consulta'], dayfirst=False,infer_datetime_format=True, errors='coerce')
# aqui convierto la columna fecha_consulta a formato fecha (datetime)
# para poder trabajarla correctamente y que no quede como texto
datos_cr[['fecha_consulta']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 1001 entries, 0 to 1019
Data columns (total 1 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   fecha_consulta  991 non-null    datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 15.6 KB


/tmp/ipykernel_19599/2819163023.py:1: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  datos_cr['fecha_consulta']= pd.to_datetime(datos_cr['fecha_consulta'], dayfirst=False,infer_datetime_format=True, errors='coerce') #que todo me lo deje año mes fecha


In [73]:
datos_cr.to_csv('datos_limpios.csv', index=False)
# finalmente guardo la base final limpia
#el index=False lo uso para que no se guarde la numeración de las filas como una columna extra en el archivo csv

INTERPRETACIÓN

Profe, al final de todo el proceso lo que hice fue limpiar la base paso a paso para que quedara más organizada y sobre todo más confiable. primero eliminé columnas que no aportaban y dejé los nombres de las variables más claros para poder trabajar mejor. después revisé duplicados y los eliminé para asegurarme de que no hubiera registros repetidos.

luego trabajé las variables categóricas como sexo, fuma y actividad_fisica, porque estaban escritas de muchas formas diferentes. ahí lo que hice fue unificar todo para que quedaran solo categorías claras, y los valores que no tenían sentido o no eran claros los dejé como nulos porque no me parecía correcto asumir cosas.

también me di cuenta que varias columnas que deberían ser numéricas estaban como texto, entonces las convertí a números para poder analizarlas bien. después de eso revisé los valores y encontré muchos fuera de rango, como edades negativas, pesos exagerados o alturas imposibles, entonces esos los pasé a nulos porque claramente eran errores en los datos.

en la presión arterial, como venía en una sola columna con dos valores, decidí separarla en presión sistólica y diastólica para poder trabajarla mejor. además, cuando una de las dos partes estaba mal o incompleta, dejé ambas como nulas porque ese dato ya no era confiable completo.

con la fecha, la convertí a formato datetime porque estaba como texto, y también calculé el imc a partir del peso y la altura. después revisé el imc y eliminé valores que no eran realistas para que quedara dentro de un rango lógico.

en cuanto a los valores nulos, decidí no rellenarlos porque esta base es tipo encuesta, entonces no tiene sentido inventar datos como la edad o el peso de una persona que no los reportó.

al final hice una verificación general revisando tipos de datos, nulos, estadísticas y cómo se veía la base, y ya todo quedó más coherente, sin valores raros y listo para análisis. básicamente con todo esto la idea es poder trabajar con datos más confiables y evitar errores en los resultados.